storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java

```diff
@@ -183,6 +205,21 @@ public boolean isFinalizedEpoch(final UnsignedLong epoch) {
     return historicalChainData.getFinalizedStateByBlockRoot(blockRoot);
   }
 
+  private BeaconState regenerateBeaconState(final BeaconState preState, final UnsignedLong slot) {
```

Comment:
Probably return `Optional<BeacaonState>` here and `Optional.empty()` instead of the `nulls and then use `flatMap` instead of `map` above.